modules/llrt_crypto/src/subtle/generate_key.rs

```diff
@@ -19,6 +23,94 @@ use super::crypto_key::{CryptoKey, CryptoKeyPair};
 static PUB_EXPONENT_1: OnceLock<BigUint> = OnceLock::new();
 static PUB_EXPONENT_2: OnceLock<BigUint> = OnceLock::new();
 
+static MANDATORY_USAGES: Lazy<HashMap<&str, HashSet<String>>> = Lazy::new(|| {
+    let mut map = HashMap::new();
+    map.insert("AES-CTR", HashSet::from([]));
+    map.insert("AES-CBC", HashSet::from([]));
+    map.insert("AES-GCM", HashSet::from([]));
+    map.insert("AES-GCM", HashSet::from([]));
+    map.insert("HMAC", HashSet::from([]));
+    map.insert("RSASSA-PKCS1-v1_5", HashSet::from(["sign".to_string()]));
+    map.insert("RSA-PSS", HashSet::from(["sign".to_string()]));
+    map.insert(
+        "RSA-OAEP",
+        HashSet::from(["decrypt".to_string(), "unwrapKey".to_string()]),
+    );
+    map.insert("ECDSA", HashSet::from(["sign".to_string()]));
+    map.insert(
+        "ECDH",
+        HashSet::from(["deriveKey".to_string(), "deriveBits".to_string()]),
+    );
+
+    map
+});
+
+static SUPPORTED_USAGES: Lazy<HashMap<&str, HashSet<String>>> = Lazy::new(|| {
+    let mut map = HashMap::new();
+    map.insert(
+        "AES-CTR",
+        HashSet::from([
+            "encrypt".to_string(),
+            "decrypt".to_string(),
+            "wrapKey".to_string(),
+            "unwrapKey".to_string(),
+        ]),
+    );
+    map.insert(
+        "AES-CBC",
+        HashSet::from([
+            "encrypt".to_string(),
+            "decrypt".to_string(),
+            "wrapKey".to_string(),
+            "unwrapKey".to_string(),
+        ]),
+    );
+    map.insert(
+        "AES-GCM",
+        HashSet::from([
+            "encrypt".to_string(),
+            "decrypt".to_string(),
+            "wrapKey".to_string(),
+            "unwrapKey".to_string(),
+        ]),
+    );
+    map.insert(
+        "AES-GCM",
+        HashSet::from(["wrapKey".to_string(), "unwrapKey".to_string()]),
+    );
+    map.insert(
+        "HMAC",
+        HashSet::from(["sign".to_string(), "verify".to_string()]),
+    );
+    map.insert(
+        "RSASSA-PKCS1-v1_5",
+        HashSet::from(["sign".to_string(), "verify".to_string()]),
+    );
+    map.insert(
+        "RSA-PSS",
+        HashSet::from(["sign".to_string(), "verify".to_string()]),
+    );
+    map.insert(
+        "RSA-OAEP",
+        HashSet::from([
+            "encrypt".to_string(),
+            "decrypt".to_string(),
+            "wrapKey".to_string(),
+            "unwrapKey".to_string(),
+        ]),
+    );
+    map.insert(
+        "ECDSA",
+        HashSet::from(["sign".to_string(), "verify".to_string()]),
+    );
+    map.insert(
+        "ECDH",
+        HashSet::from(["deriveKey".to_string(), "deriveBits".to_string()]),
+    );
+
+    map
+});
+
```

Comment:
This can be heavily optimized by using static arrays instead of Lazy HashMaps. It's almost always better to use an array when dealing with few items especially if the keys are strings with short length. Also when having the same value twice we can directly work with references:

```rust
static SYMMETRIC_USAGES: &[&str] = &["encrypt", "decrypt", "wrapKey", "unwrapKey"];
static SIGNATURE_USAGES: &[&str] = &["sign", "verify"];
static EMPTY_USAGES: &[&str] = &[];
static SIGN_USAGES: &[&str] = &["sign"];
static RSA_OAEP_USAGES: &[&str] = &["decrypt", "unwrapKey"];
static ECDH_USAGES: &[&str] = &["deriveKey", "deriveBits"];

static SUPPORTED_USAGES_ARRAY: &[(&str, &[&str])] = &[
    ("AES-CTR", SYMMETRIC_USAGES),
    ("AES-CBC", SYMMETRIC_USAGES),
    ("AES-GCM", SYMMETRIC_USAGES),
    ("RSA-OAEP", SYMMETRIC_USAGES),
    ("HMAC", SIGNATURE_USAGES),
    ("RSASSA-PKCS1-v1_5", SIGNATURE_USAGES),
    ("RSA-PSS", SIGNATURE_USAGES),
    ("ECDSA", SIGNATURE_USAGES),
    ("ECDH", &["deriveKey", "deriveBits"]),
];

static MANDATORY_USAGES_ARRAY: &[(&str, &[&str])] = &[
    ("AES-CTR", EMPTY_USAGES),
    ("AES-CBC", EMPTY_USAGES),
    ("AES-GCM", EMPTY_USAGES),
    ("HMAC", EMPTY_USAGES),
    ("RSASSA-PKCS1-v1_5", SIGN_USAGES),
    ("RSA-PSS", SIGN_USAGES),
    ("ECDSA", SIGN_USAGES),
    ("RSA-OAEP", RSA_OAEP_USAGES),
    ("ECDH", ECDH_USAGES),
];
```

We can also optimize the usages check by just checking if we're using a non supported value or missing a mandatory, for example:
```rust
fn process_usages(
    mandatory_usages: &[&str],
    supported_usages: &[&str],
    key_usages_set: HashSet<String>,
) -> Result<HashSet<String>> {
    //early check for mandatory usages
    if key_usages_set.len() != mandatory_usages.len() {
        return Err(...);
    }

    //check all usages are supported first
    if !key_usages_set
        .iter()
        .all(|u| supported_usages.contains(u.as_str()))
    {
         return Err(...);
    }

    //check mandatory usages are present
    if !mandatory_usages.iter().all(|m| key_usages_set.contains(*m)) {
        return Err(...);
    }

    //all checks passed
    Ok(key_usages_set)
}

```
There is no clones, filters or duplication of data here :)

Reply:
This seems like it could be significantly optimized! However, it would require separating public_usages and private_usages.

Reply:
This was just psudo code, maybe you can apply the same concept for both at the same time? Are `public_or_secret_usages` not checked for supported but only filtered by mandatory? What's the difference between `public_or_secret_usages` and `private_usages`?

Reply:
I could not find any [specification](https://w3c.github.io/webcrypto/#keypair) that separates `usages` for privateKey and publicKey when generateKey returns CryptoKeyPair.

After checking the behavior of bun, I found that it follows the following rules, so I am using them as reference.

1. If `MANDATORY_USAGES` is defined, at least one of them must be specified (an error occurs if none are specified)
2. Only those defined in `SUPPORTED_USAGES` must be specified (an error occurs if anything other than `SUPPORTED_USAGES` is specified)
3. Only privateKey `usages` included in `MANDATORY_USAGES` are passed.
4. Only publicKey `usages` not included in `MANDATORY_USAGES` are passed.

```javascript
// subtleGenerateKey.js
const key = await crypto.subtle
  .generateKey(
    {
      name: "ECDSA",
      namedCurve: "P-256",
    },
    true,
    ["sign", "verify"]
  );

console.log(key);
```
```
% bun subtleGenerateKey.js 
{
  privateKey: CryptoKey {
    type: "private",
    extractable: true,
    algorithm: {
      name: "ECDSA",
      namedCurve: "P-256",
    },
    usages: [ "sign" ],
  },
  publicKey: CryptoKey {
    type: "public",
    extractable: true,
    algorithm: {
      name: "ECDSA",
      namedCurve: "P-256",
    },
    usages: [ "verify" ],
  },
}
```